File: Beecrowd_Parte4/Presentes_Grandes_2720.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int id;
    long long volume;
} Present;

int compare(const void *a, const void *b) {
    Present *presentA = (Present *)a;
    Present *presentB = (Present *)b;
    if (presentB->volume != presentA->volume) {
        return presentB->volume - presentA->volume;
    } else {
        return presentA->id - presentB->id;
    }
}

int compareInt(const void *a, const void *b) {
    return (*(int*)a - *(int*)b);
}

int main() {
    int T, N, K, H, W, L;
    scanf("%d", &T);

    for (int t = 0; t < T; t++) {
        scanf("%d %d", &N, &K);
        Present presents[N];

        for (int i = 0; i < N; i++) {
            scanf("%d %d %d %d", &presents[i].id, &H, &W, &L);
            presents[i].volume = (long long)H * W * L;
        }

        qsort(presents, N, sizeof(Present), compare);

        int ids[K];
        for (int i = 0; i < K; i++) {
            ids[i] = presents[i].id;
        }

        qsort(ids, K, sizeof(int), compareInt);

        for (int i = 0; i < K; i++) {
            if (i > 0) {
                printf(" ");
            }
            printf("%d", ids[i]);
        }

        printf("\n");
    }

    return 0;
}

File: Beecrowd_Parte4/Presentes_Grandes_2720.java
import java.util.Arrays;
import java.util.Scanner;

class Present {
    int id;
    long volume;

    Present(int id, long volume) {
        this.id = id;
        this.volume = volume;
    }
}

public class Presentes_Grandes_2720 {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        int T = scanner.nextInt();

        for (int t = 0; t < T; t++) {
            int N = scanner.nextInt();
            int K = scanner.nextInt();
            Present[] presents = new Present[N];

            for (int i = 0; i < N; i++) {
                int id = scanner.nextInt();
                long volume = 1L * scanner.nextInt() * scanner.nextInt() * scanner.nextInt();
                presents[i] = new Present(id, volume);
            }

            Arrays.sort(presents, (a, b) -> {
                if (b.volume != a.volume) {
                    return Long.compare(b.volume, a.volume);
                } else {
                    return Integer.compare(a.id, b.id);
                }
            });

            int[] ids = new int[K];
            for (int i = 0; i < K; i++) {
                ids[i] = presents[i].id;
            }

            Arrays.sort(ids);

            for (int i = 0; i < K; i++) {
                if (i > 0) {
                    System.out.print(" ");
                }
                System.out.print(ids[i]);
            }

            System.out.println();
        }

        scanner.close();
    }
}